doc/specs/#885 - winrt Terminal Settings.md

```diff
@@ -0,0 +1,205 @@
+This API will look something like this:
+```c++
+runtimeclass AppSettings
+{
+    AppSettings();
+
+    // Load a settings file, and layer those changes on top of the existing AppSettings
+    void LayerSettings(String path);
+
+    // Create a copy of the existing AppSettings
+    AppSettings Clone();
+    
+    // Compares object to "source" and applies changes to
+    // the settings file at "outPath"
+    void ApplyChanges(String outPath, AppSettings source);
+}
+```
+
+#### TerminalApp: Loading and Reloading Changes
+
+TerminalApp will construct and reference an `AppSettings settings` using a shared smart pointer as follows:
+- TerminalApp will have a global reference to "defaults.json" and "settings.json" filepaths as `constexpr`
+- construct an `AppSettings` using the default constructor
+- `settings.LayerSettings("defaults.json")`
+- `settings.LayerSettings("settings.json")`
+
+**NOTE:** This model allows us to layer even more settings files on top of the existing Terminal Settings
+ Model, if so desired. This could be helpful when importing additional settings files from an external location
+ such as a marketplace.
+
+When TerminalApp detects a change to "settings.json", it'll repeat the steps above. We could cache the result from
+ `settings.LayerSettings("defaults.json")` to improve performance.
+
+Additionally, TerminalApp would reference `settings` for the following functionality...
+- `Profiles` to populate the dropdown menu
+- `Keybindings` to detect active key bindings and which actions they run
+- `Commands` to populate the Command Palette
+- `Warnings` to show a serialization warning, if any
+
+#### TerminalControl: Acquiring and Applying the Settings
+
+At the time of writing this spec, TerminalApp constructs `TerminalControl.TerminalSettings` WinRT objects
+ to expose `IControlSettings` and `ICoreSettings` to any hosted terminals. In moving `IControlSettings`
+ and `ICoreSettings` down to the TerminalControl layer, TerminalApp can now have better control over
+ how to expose relevant settings to a TerminalControl instance.
+
+`TerminalSettings` (which implements `IControlSettings` and `ICoreSettings`) will be moved to TerminalApp and act as a bridge that queries `AppSettings`. Thus,
+ when TermControl requests `IControlSettings` data, `TerminalSettings` responds by extracting the
+ relevant information from `AppSettings`. TerminalCore operates the same way with `ICoreSettings`.
+
+#### Settings UI: Modifying and Applying the Settings
+
+The Settings UI will also have a reference to the `AppSettings settings` from TerminalApp using a shared smart pointer
+as `settingsSource`. When the Settings UI is opened up, the Settings UI will also have its own `AppSettings settingsClone`
+that is a clone of TerminalApp's `AppSettings`.
+```c++
+settingsClone = settingsSource.Clone()
+```
+
+As the user navigates the Settings UI, the relevant contents of `settingsSource` will be retrieved and presented.

```

Comment:
I think it would be better to data-bind directly to `settingsClone` instead of retrieving and presenting values from `settingsSource`.

By using data binding, you also avoid having "XAML" "hold on to the changes". The changes can be staged _live_ into the cloned settings object.

If you're going to make "save/apply" commit the settings _into the clone_, there's no point in having the clone.

The clone is what we're using to _do_ save/apply.